phoenix-core/src/main/java/org/apache/phoenix/hbase/index/IndexRegionObserver.java

```diff
@@ -227,7 +228,7 @@ public static class BatchMutateContext {
       // The collection of candidate index mutations that will be applied after the data table mutations
       private ListMultimap<HTableInterfaceReference, Pair<Mutation, byte[]>> indexUpdates;
       private List<RowLock> rowLocks = Lists.newArrayListWithExpectedSize(QueryServicesOptions.DEFAULT_MUTATE_BATCH_SIZE);
-      private HashSet<ImmutableBytesPtr> rowsToLock = new HashSet<>();
+      private Set<ImmutableBytesPtr> rowsToLock = new TreeSet<>();
```

Comment:
Also, it would be good to put a comment above the TreeSet constructor mentioning that it's important that the set be sorted to avoid deadlocks. (Just in case someone comes across it later and thinks about making it more "efficient" by switching back to HashSet)

Reply:
Sounds good. +1 after this change is done.

Reply:
>It only matters that all threads use a consistent sort order to acquire locks, not that that consistent order matches any particular order (such as the HBase row key comparator), correct?

That is right. 

>Also, it would be good to put a comment above the TreeSet constructor

Done
@virajjasani @gjacoby126 FYI